scripts/globals/garrison.lua

```diff
@@ -513,165 +178,318 @@ local spawnNPCs = function(npc)
+    if mob == nil then
+        return nil
+    end
+
+    xi.garrison.addLevelCap(mob, zoneData.levelCap)
+    mob:setRoamFlags(xi.roamFlag.SCRIPTED)
+    table.insert(zoneData.mobs, mobID)
+
+    -- Death listener for tracking win/lose condition
+    mob:addListener("DEATH", "GARRISON_MOB_DEATH", function(mobArg)
+        zoneData.deadMobCount = zoneData.deadMobCount + 1
+    end)
+
+    -- A wave is considered complete when all mobs are done despawning
+    -- and not just dead. This matters a lot because of spawn timings.
+    -- I.e: If mob A dies on wave 1, and another instance of mob A is supposed
+    -- to spawn on wave 2, it will not spawn as long as the previous mob is still
+    -- despawning
+    -- For this reason, we track both death and despawn as separate events
+    mob:addListener("DESPAWN", "GARRISON_MOB_DESPAWN", function(mobArg)
+        zoneData.despawnedMobCount = zoneData.despawnedMobCount + 1
+    end)
+
+    return mob
+end
+
+-- Given a starting mobID, return the list of randomly selected
+-- mob ids. The amount of mobs selected is determined by numMobs.
+-- The ids in the given excludedMobs table will not be included in the result.
+-- This method assumes that the mob pool is composed by mobIDs
+-- that are sequential between firstMobID and lastMobID
+-- e.g: If firstMobId = 1, lastMobID = 4 and numMobs is 2,
+-- Then 2 ids randomly selected between { 1, 2, 3, 4 } will be returned
+-- without repetitions.
+xi.garrison.pickMobsFromPool = function(firstMobID, lastMobID, numMobs, excludedMobIDs)
+    -- unfiltered pool, from first to last mob id (inclusive)
+    local unfilteredPool = utils.range(firstMobID, lastMobID)
+
+    -- filter the pool, removing excludedMobIDs
+    local pool = {}
+    local excludedSet = set(excludedMobIDs)
+    for _, v in ipairs(unfilteredPool) do
+        if not excludedSet[v] then
+            table.insert(pool, v)
+        end
+    end
+
+    -- validate input
+    local mobs = {}
+    if numMobs > #pool then
+        printf("[warning] pickMobsFromPool called with numMobs > mobIds. Num Mobs: %i. Pool size: %i", numMobs, #pool)
+        numMobs = #pool
+    end
+
+    if numMobs <= 0 then
+        printf("[error] Invalid numMobs picked. Should be > 0.")
+        return {}
+    end
+
+    -- Now we can apply a common algorithm used to "shuffle a deck of cards"
+    for i = 1, numMobs do
+        -- Pick random index from J to pool end. Add the picked element to result
+        local pickedIndex = math.random(i, #pool)
+        table.insert(mobs, pool[pickedIndex])
+
+        -- Now swap the picked element with the first element of the array.
+        -- This effectively makes the picked element not elegible for future picks.
+        pool[pickedIndex], pool[i] = pool[i], pool[pickedIndex]
+    end
+
+    return mobs
+end
+
 -----------------------------------
 -- Main Functions
 -----------------------------------
 
 xi.garrison.tick = nil -- Prototype
 xi.garrison.tick = function(npc)
```

Comment:
one thing that makes me pretty nervous about this code is that if there is any kind of exception during tick, npcs are never cleaned up. What's a good approach to solve this? I was thinking of adding some npc:timer to check if everything is alright at a certain ratio, but I'm unsure if that is overkill. Another option would be to use pcall() and catch any errors thrown by tick(). That does have a perf issue but imo it's pretty clean

Reply:
I think a periodic timer attached to the NPC is fine. You can install it just as a garrison in a zone is about to start, and only remove it (not renew it) when you're in a known good and clean state. It doesn't need to be ticking all the time, forever.

Reply:
I am going to make a separate PR for the timer since the bulk functionality is there and fully tested.